cryptol-saw-core/src/Verifier/SAW/Cryptol.hs

```diff
@@ -1953,38 +2031,518 @@ exportRecordValue fields v =
+
   where
-    genConstr :: Env -> NominalType -> IO Env
-    genConstr env nt = do
+
+    updateEnvForNominal :: Env -> NominalType -> IO Env
+    updateEnvForNominal env nt = do
+      let kinds = map C.tpKind (C.ntParams nt)
+      unless (all (`elem` [C.KType, C.KNum]) kinds) $
+        panic "genCodeForNominalTypes"
+          [ "type parameters for nominal types must each have kind * or #:"
+          , show kinds
+          ]
+
+      ns <- newDefsForNominal env nt
       let conTs = C.nominalTypeConTypes nt
-      constrs <- forM (nominalConstrs nt) $ \(x,e) ->
-        do e' <- importExpr sc env e
-           pure (x,(e',0))
+          constrs = map
+                      (\(x,e) -> (x,(e,0)))  -- the De Bruijn 'magic'
+                      ns
       let env' = env { envE = foldr (uncurry Map.insert) (envE env) constrs
                      , envC = foldr (uncurry Map.insert) (envC env) conTs
                      }
+        -- NOTE: the Cryptol schemas for the Struct & Enum constructors get added to
+        --       the Cryptol environment.
+
       return env'
 
-    nominalConstrs :: NominalType -> [(C.Name,C.Expr)]
-    nominalConstrs nt =
+    -- | Create functions/constructors for different 'NominalType's.
+    newDefsForNominal ::
+      HasCallStack => Env -> NominalType -> IO [(C.Name,Term)]
+    newDefsForNominal env nt =
       case C.ntDef nt of
-        C.Struct fs ->
-          let recTy = C.TRec (C.ntFields fs)
-              fn    = C.EAbs paramName recTy (C.EVar paramName)
-              con   = C.ntConName fs
-              paramName = C.asLocal C.NSValue con
-           in [(con, foldr tFn fn (C.ntParams nt))]
-        C.Abstract -> []
-        C.Enum {} -> error "genNominalConstrurctors: `enum` is not yet supported"
-      where
+        C.Abstract  -> return []
+
+        C.Enum x    -> genCodeForEnum sc env nt x
+
+        C.Struct fs -> do
+            let recTy      = C.TRec (C.ntFields fs)
+                conNm      = C.ntConName fs
+                paramName  = C.asLocal C.NSValue conNm
+                             -- NOTE: We use name of constructor as the
+                             -- name of the constructor argument! Thus we can
+                             -- avoid need for a new unique name.
+                             -- FIXME: this doesn't seem foolproof!
+                fn         = C.EAbs paramName recTy (C.EVar paramName)
+                fnWithTAbs = foldr C.ETAbs fn (C.ntParams nt)
+            e <- importExpr sc env fnWithTAbs
+            return [(conNm, e)]
+
+
+-- | genCodeForEnum ... - called when we see an "enum" definition in the Cryptol module.
+--    - This action does two things
+--       1. Returns the names & definitions of the constructors of the enum.
+--          This fit with the code for other nominals, needed because
+--          the "rest" of cryptol code to be translated needs to see the
+--          onstructors in the Cryptol environments.
```

Comment:
```suggestion
--       1. Returns the names & definitions of the constructors of the enum.
--          This fits with the code for other nominals, needed because
--          the "rest" of cryptol code to be translated needs to see the
--          constructors in the Cryptol environments.
```